codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2ConnectionEncoder.java

```diff
@@ -464,4 +439,140 @@ public Http2Settings pollSentSettings() {
     public Configuration configuration() {
         return frameWriter.configuration();
     }
+
+    /**
+     * Wrap a DATA frame so it can be written subject to flow-control. Note that this implementation assumes it
+     * only writes padding once for the entire payload as opposed to writing it once per-frame. This makes the
+     * {@link #size} calculation deterministic thereby greatly simplifying the implementation.
+     * <p>
+     * If frame-splitting is required to fit within max-frame-size and flow-control constraints we ensure that
+     * the passed promise is not completed until last frame write.
+     * </p>
+     */
+    private final class FlowControlledData extends FlowControlledPayloadBase {
+        private ByteBuf data;
+        private int size;
+
+        private FlowControlledData(ChannelHandlerContext ctx, Http2Stream stream, ByteBuf data, int padding,
+                                    boolean endOfStream, ChannelPromise promise) {
+            super(ctx, stream, padding, endOfStream, promise);
+            this.data = data;
+            size = data.readableBytes() + padding;
+        }
+
+        @Override
+        public int size() {
+            return size;
+        }
+
+        @Override
+        public void error(Throwable cause) {
+            ReferenceCountUtil.safeRelease(data.release());
+            data = null;
+            size = 0;
+            promise.tryFailure(cause);
+        }
+
+        @Override
+        public boolean write(int allowedBytes) {
+            if (data == null) {
+                return false;
+            }
+            if (allowedBytes == 0 && size() != 0) {
+                // No point writing an empty DATA frame, wait for a bigger allowance.
+                return false;
+            }
+            int maxFrameSize = frameWriter().configuration().frameSizePolicy().maxFrameSize();
+            try {
+                int bytesWritten = 0;
+                do {
+                    int allowedFrameSize = Math.min(maxFrameSize, allowedBytes - bytesWritten);
+                    ByteBuf toWrite;
+                    // Let data consume the frame before padding.
+                    int writeableData = data.readableBytes();
+                    if (writeableData > allowedFrameSize) {
+                        writeableData = allowedFrameSize;
+                        toWrite = data.readSlice(writeableData).retain();
+                    } else {
+                        // We're going to write the full buffer which will cause it to be released, for subsequent
+                        // writes just use empty buffer to avoid over-releasing.
+                        toWrite = data;
+                        data = Unpooled.EMPTY_BUFFER;
```

Comment:
Maybe add a comment that you're setting to EMPTY_BUFFER here so that this method continues to work if we need to send more padding.